File: XMLCheck/Handler/CheckRulesXML01/Check_NGAY_SINH.cs
using XMLCheck.Handler.CheckRulesXML1;
using XMLCheck.Models;
using XMLCheck.Services;

namespace XMLCheck.Handler.CheckRulesXML01
{
    public class Check_NGAY_SINH:Xml1RuleBase
    {
        public override string Key => "NGAY_SINH";

        public override ErrorDetails? Check(Xml1Model model, HoSoContext context)
        {

            // 1. Không được để trống
            if (model.NGAY_SINH == null)
            {
                return Error("NGAY_SINH không được để trống");
            }

            var ngaySinh = model.NGAY_SINH.Value;
            var now = DateTime.Now;

            // 2. Không được lớn hơn ngày hiện tại
            if (ngaySinh > now)
            {
                return Error("NGAY_SINH không được lớn hơn ngày hiện tại");
            }

            // 3. Không được nhỏ hơn 140 năm
            if (ngaySinh < now.AddYears(-140))
            {
                return Error("NGAY_SINH không được nhỏ hơn 140 năm so với thời gian hiện tại");
            }

            /*
             * ⚠️ Các rule sau KHÔNG thể kiểm tra ở tầng này:
             * - Không đúng định dạng yyyyMMddHHmmss
             * - Sai định dạng
             *
             * Vì:
             * NGAY_SINH đã được parse sang DateTime? ở Parser
             * Nếu sai format → đã thành null → rơi vào rule (1)
             */

            return null;
        }
    }
}

File: XMLValidator_v2/src/parsers/rules/xml1/Check_NGAY_SINH.ts
import { Xml1RuleBase } from './xml1RuleBase';
import { Xml1Model } from '../../../interfaces/xml-models.interface';
import { HoSoContext } from '../../../services/hoso-context';
import { ErrorDetails } from '../../../dto/xml-check.dto';

export class Check_NGAY_SINH extends Xml1RuleBase {
  constructor() {
    super({
      ruleId: 'RULE_XML1_NGAY_SINH_01',
      xmlType: 'XML1',
      field: 'NGAY_SINH',
      severity: 'ERROR',
      description: 'Kiểm tra quy định cho trường NGAY_SINH',
      errorMessage: 'NGAY_SINH không được để trống',
      reference: 'QĐ 3176/QĐ-BYT',
    });
  }

  

  public check(model: Xml1Model, context: HoSoContext): ErrorDetails | null {
    if (!model.NGAY_SINH) {
      return this.error('NGAY_SINH không được để trống');
    }

    const ngaySinh = model.NGAY_SINH;
    const now = new Date();

    if (ngaySinh > now) {
      return this.error('NGAY_SINH không được lớn hơn ngày hiện tại');
    }

    const minDate = new Date();
    minDate.setFullYear(now.getFullYear() - 140);
    if (ngaySinh < minDate) {
      return this.error('NGAY_SINH không được nhỏ hơn 140 năm so với thời gian hiện tại');
    }

    return null;
  }
}
